docs/cudf/source/developer_guide/testing.md

```diff
@@ -0,0 +1,143 @@
+
+When it comes to parametrizing tests, the two main options are fixtures and `mark.parametrize`.
+By virtue of being functions, fixtures are both more verbose and more self-documenting.
+Fixtures also have the significant benefit of being constructed lazily,
+whereas parametrizations are constructed at test collection time.
+
+In general, these approaches are applicable to parametrizations of different complexity.
+For the purpose of this discussion,
+we define a parametrization as "simple" if it is composed of a list (possibly nested) of primitive objects.
+Examples include a list of integers or a list of list of strings.
+This _does not_ include e.g. cuDF or pandas objects.
+
+With that in mind, here are some ground rules for how to parametrize.
+
+Use `pytest.mark.parametrize` when:
+- One test must be run on many inputs and those inputs are simple to construct.
+
+Use fixtures when:
+- One or more tests must be run on the same set of inputs,
+  and all of those inputs can be constructed with simple parametrizations.
+  In practice, that means that it is acceptable to use a fixture like this:
+  ```python
+      @pytest.fixture(params=['a', 'b'])
+      def foo(request):
+          if request.param == 'a':
+              # Some complex initialization
+          elif request.param == 'b':
+              # Some other complex initialization
+  ```
+  In other words, the construction of the fixture may be complex,
+  as long as the parametrization of that construction is simple.
+- One or more tests must be run on the same set of inputs,
+  and at least one of those inputs requires complex parametrizations.
+  In this case, the parametrization of a fixture should be decomposed
+  by using fixtures that depend on other fixtures.
+  ```python
+      @pytest.fixture(params=['a', 'b'])
+      def foo(request):
+          if request.param == 'a':
+              # Some complex initialization
+          elif request.param == 'b':
+              # Some other complex initialization
+
+      def bar(foo):
+         # do something with foo like initialize a cudf object.
+  ```
+
+#### Complex parametrizations
+
+The lists above document common use cases.
+However, more complex cases may arise.
+One of the most common alternatives is where, given a set of test cases,
+different tests need to run on different subsets with a nonempty intersection.
+Fixtures and parametrization are only capable of handling the Cartesian product of parameters,
+i.e. "run this test for all values of `a` and all values of `b`".
+
+There are multiple potential solutions to this problem.
+One possibility is to encapsulate common test logic a helper function,
+then call it from multiple `test_*` functions that construct the necessary inputs.
+Another possibility is to use functions rather than fixtures to construct inputs, allowing for more flexible input construction:
+```python
+def get_values(predicate):
+    values = [...]
```

Comment:
Real-ish input values can help illustrate this.
```suggestion
    values = list(range(10))
```

Reply:
No need for the list, but yes good idea.